File: apps/workflow-engine/src/nodes/SplitInBatches.ts
import type { NodeDefinition } from '../schemas/workflow.js';
import type {
  ExecutionContext,
  NodeData,
  NodeExecutionResult,
} from '../engine/types.js';
import type { INodeTypeDescription } from '../engine/nodeSchema.js';
import { BaseNode } from './BaseNode.js';

interface SplitState {
  items: NodeData[];
  batchIndex: number;
  totalItems: number;
  [key: string]: unknown; // Index signature for compatibility with Record<string, unknown>
}

/**
 * SplitInBatches node - process arrays in chunks with looping
 *
 * Outputs:
 * - loop: Current batch (connects to nodes that process the batch)
 * - done: Final output when all batches processed
 *
 * The "loop" output should connect to processing nodes, which eventually
 * connect back to this SplitInBatches node to continue with next batch.
 */
export class SplitInBatchesNode extends BaseNode {
  readonly type = 'SplitInBatches';
  readonly description = 'Process items in batches with looping';

  static readonly nodeDescription: INodeTypeDescription = {
    name: 'SplitInBatches',
    displayName: 'Split In Batches',
    icon: 'fa:th-large',
    description: 'Process items in batches with looping',
    group: ['flow'],

    outputs: [
      { name: 'loop', displayName: 'Loop' },
      { name: 'done', displayName: 'Done' },
    ],

    properties: [
      {
        displayName: 'Batch Size',
        name: 'batchSize', // ← matches getParameter('batchSize')
        type: 'number',
        default: 10,
        typeOptions: { minValue: 1 },
        description: 'Number of items to process in each batch',
      },
    ],
  };

  async execute(
    context: ExecutionContext,
    nodeDefinition: NodeDefinition,
    inputData: NodeData[]
  ): Promise<NodeExecutionResult> {
    const batchSize = this.getParameter<number>(nodeDefinition, 'batchSize', 10);
    const stateKey = nodeDefinition.name;

    // Get or initialize state
    let state = context.nodeInternalState.get(stateKey) as SplitState | undefined;

    if (!state) {
      // First run - initialize with all input items
      state = {
        items: inputData,
        batchIndex: 0,
        totalItems: inputData.length,
      };
      context.nodeInternalState.set(stateKey, state);
    }

    const startIdx = state.batchIndex;
    const endIdx = Math.min(startIdx + batchSize, state.items.length);
    const batch = state.items.slice(startIdx, endIdx);

    if (batch.length === 0) {
      // No more items - we're done
      context.nodeInternalState.delete(stateKey);
      return this.outputs({
        loop: null,
        done: [
          {
            json: {
              totalProcessed: state.totalItems,
              batchesProcessed: Math.ceil(state.totalItems / batchSize),
            },
          },
        ],
      });
    }

    // Update state for next iteration
    state.batchIndex = endIdx;
    context.nodeInternalState.set(stateKey, state);

    // Return current batch on "loop" output
    return this.outputs({
      loop: batch,
      done: null,
    });
  }
}

File: apps/workflow-engine/src/nodes/split_in_batches.py
"""SplitInBatches node - process array items in batches."""

from __future__ import annotations

from typing import Any, TYPE_CHECKING

from .base import (
    BaseNode,
    NodeTypeDescription,
    NodeInputDefinition,
    NodeOutputDefinition,
    NodeProperty,
)

if TYPE_CHECKING:
    from ..engine.types import ExecutionContext, NodeData, NodeDefinition, NodeExecutionResult


class SplitInBatchesNode(BaseNode):
    """SplitInBatches node - process array items in batches with loop support."""

    node_description = NodeTypeDescription(
        name="SplitInBatches",
        display_name="Split In Batches",
        description="Process array items in batches with loop support",
        icon="fa:layer-group",
        group=["flow"],
        inputs=[NodeInputDefinition(name="main", display_name="Input")],
        outputs=[
            NodeOutputDefinition(
                name="loop",
                display_name="Loop",
                schema={"type": "unknown", "description": "Current batch items"},
            ),
            NodeOutputDefinition(
                name="done",
                display_name="Done",
                schema={"type": "unknown", "description": "All items after processing"},
            ),
        ],
        properties=[
            NodeProperty(
                display_name="Batch Size",
                name="batchSize",
                type="number",
                default=10,
                description="Number of items per batch",
            ),
            NodeProperty(
                display_name="Reset",
                name="reset",
                type="boolean",
                default=False,
                description="Reset batch state (start from beginning)",
            ),
        ],
    )

    @property
    def type(self) -> str:
        return "SplitInBatches"

    @property
    def description(self) -> str:
        return "Process array items in batches with loop support"

    async def execute(
        self,
        context: ExecutionContext,
        node_definition: NodeDefinition,
        input_data: list[NodeData],
    ) -> NodeExecutionResult:
        from ..engine.types import NodeData

        batch_size = self.get_parameter(node_definition, "batchSize", 10)
        reset = self.get_parameter(node_definition, "reset", False)

        # Get or initialize internal state
        state_key = node_definition.name
        state: dict[str, Any] = context.node_internal_state.get(state_key, {})

        if reset or "items" not in state:
            # First run or reset - store all items and start at index 0
            all_items = [item.json for item in input_data]
            state = {
                "items": all_items,
                "currentIndex": 0,
                "processedItems": [],
            }

        items = state["items"]
        current_index = state["currentIndex"]

        # Get the next batch
        batch_end = min(current_index + batch_size, len(items))
        batch = items[current_index:batch_end]

        if not batch:
            # No more items - output done
            context.node_internal_state.pop(state_key, None)
            all_processed = state.get("processedItems", items)
            return self.outputs({
                "loop": None,
                "done": [NodeData(json=item) for item in all_processed],
            })

        # Update state for next iteration
        state["currentIndex"] = batch_end
        state["processedItems"].extend(batch)
        context.node_internal_state[state_key] = state

        # If there are more items, output to loop; otherwise mark as done
        has_more = batch_end < len(items)

        return self.outputs({
            "loop": [NodeData(json=item) for item in batch] if has_more else None,
            "done": None if has_more else [NodeData(json=item) for item in state["processedItems"]],
        })
